scheduler/reconcile.go

```diff
@@ -0,0 +1,559 @@
+		a.deploymentFailed = deployment.Status == structs.DeploymentStatusFailed
+	}
+
+	return a
+}
+
+// Compute reconciles the existing cluster state and returns the set of changes
+// required to converge the job spec and state
+func (a *allocReconciler) Compute() *reconcileResults {
+	// Create the allocation matrix
+	m := newAllocMatrix(a.job, a.existingAllocs)
+
+	// Handle stopping unneeded deployments
+	a.cancelDeployments()
+
+	// If we are just stopping a job we do not need to do anything more than
+	// stopping all running allocs
+	if a.job.Stopped() {
+		a.handleStop(m)
+		return a.result
+	}
+
+	// Reconcile each group
+	for group, as := range m {
+		a.computeGroup(group, as)
+	}
+
+	return a.result
+}
+
+// cancelDeployments cancels any deployment that is not needed
+func (a *allocReconciler) cancelDeployments() {
+	// If the job is stopped and there is a non-terminal deployment, cancel it
+	if a.job.Stopped() {
+		if a.deployment != nil && a.deployment.Active() {
+			a.result.deploymentUpdates = append(a.result.deploymentUpdates, &structs.DeploymentStatusUpdate{
+				DeploymentID:      a.deployment.ID,
+				Status:            structs.DeploymentStatusCancelled,
+				StatusDescription: structs.DeploymentStatusDescriptionStoppedJob,
+			})
+		}
+
+		// Nothing else to do
+		return
+	}
+
+	// Check if the deployment is referencing an older job and cancel it
+	if d := a.deployment; d != nil {
+		if d.Active() && (d.JobCreateIndex != a.job.CreateIndex || d.JobModifyIndex != a.job.JobModifyIndex) {
+			a.result.deploymentUpdates = append(a.result.deploymentUpdates, &structs.DeploymentStatusUpdate{
+				DeploymentID:      a.deployment.ID,
+				Status:            structs.DeploymentStatusCancelled,
+				StatusDescription: structs.DeploymentStatusDescriptionNewerJob,
+			})
+			a.deployment = nil
+		}
+	}
+}
+
+// handleStop marks all allocations to be stopped, handling the lost case
+func (a *allocReconciler) handleStop(m allocMatrix) {
+	for group, as := range m {
+		untainted, migrate, lost := as.filterByTainted(a.taintedNodes)
+		a.markStop(untainted, "", allocNotNeeded)
+		a.markStop(migrate, "", allocNotNeeded)
+		a.markStop(lost, structs.AllocClientStatusLost, allocLost)
+		desiredChanges := new(structs.DesiredUpdates)
+		desiredChanges.Stop = uint64(len(as))
+		a.result.desiredTGUpdates[group] = desiredChanges
+	}
+}
+
+// markStop is a helper for marking a set of allocation for stop with a
+// particular client status and description.
+func (a *allocReconciler) markStop(allocs allocSet, clientStatus, statusDescription string) {
+	for _, alloc := range allocs {
+		a.result.stop = append(a.result.stop, allocStopResult{
+			alloc:             alloc,
+			clientStatus:      clientStatus,
+			statusDescription: statusDescription,
+		})
+	}
+}
+
+// computeGroup reconciles state for a particular task group.
+func (a *allocReconciler) computeGroup(group string, all allocSet) {
```

Comment:
It would be nice to break down this function a bit more, it's hard to maintain all the context to follow